consensus/hotstuff/timeoutcollector/timeout_processor.go

```diff
@@ -0,0 +1,298 @@
+package timeoutcollector
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/signature"
+)
+
+// accumulatedWeightTracker tracks one-time event of reaching required weight
+// Uses atomic flag to guarantee concurrency safety.
+type accumulatedWeightTracker struct {
+	minRequiredWeight uint64
+	done              atomic.Bool
+}
+
+func (t *accumulatedWeightTracker) Done() bool {
+	return t.done.Load()
+}
+
+// Track checks if required threshold was reached as one-time event and
+// returns true whenever it's reached.
+func (t *accumulatedWeightTracker) Track(weight uint64) bool {
+	if weight < t.minRequiredWeight {
+		return false
+	}
+	if t.done.CAS(false, true) {
+		return true
+	}
+	return false
+}
+
+// NewestQCTracker is a helper structure which keeps track of the highest QC(by view)
+// in concurrency safe way.
+type NewestQCTracker struct {
+	lock     sync.RWMutex
+	newestQC *flow.QuorumCertificate
```

Comment:
I am not sure if it will work (but it might be an interesting experiment nevertheless):
* I think we might be able to use [`atomic.Value`](https://pkg.go.dev/sync/atomic#Value) here . It implements `CompareAndSwap`, which could be sufficient for our algorithm. My hope is that an atomic would be more performant than a mutex and less prone to congestion.  
* Only downside of using `atomic.Value` is that it doesn't handle `nil`. So we would need to initialize with a QC. While we always have a QC, it might bloat the code too much to add it to all the constructors.

Reply:
https://github.com/onflow/flow-go/pull/2620/commits/e9d18cc4d891953b9a3b747d7b439707429dfac7